README.md
cov
=======

[![Dependency Status](https://gemnasium.com/fengmk2/cov.png)](https://gemnasium.com/fengmk2/cov)

[![NPM](https://nodei.co/npm/cov.png?downloads=true&stars=true)](https://nodei.co/npm/cov/)

Mocha Coverage report store on CDN.

## Install

```bash
$ npm install cov --registry=http://registry.cnpmjs.org
```

## Coverage Config

Add the key `"cov"` to the `"config"` key in your package.json file.
Under that key you can add any of the following properties:

```json
"config": {
  "cov": {
    "threshold": 100
  }
}
```

## Usage

@see [Makefile](https://github.com/fengmk2/cov/blob/master/Makefile)

```bash
$ make test MOCHA_OPTS='--require blanket' REPORTER=html-cov | ./node_modules/cov/bin/cov
```

## License

(The MIT License)

Copyright (c) 2013 fengmk2 &lt;fengmk2@gmail.com&gt;

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.md

0.1.3 / 2014-04-09
==================

 * ignore upload error

0.1.2 / 2014-03-14 
==================

  * add node version to cdn dir

0.1.1 / 2014-03-04 
==================

  * fix index page

0.1.0 / 2014-03-04 
==================

  * support istanbul coverage

0.0.2 / 2014-01-13 
==================

  * support config["travis-cov"]

0.0.1 / 2013-12-20 
==================

  * first commit


lib/stdin.js
/**!
 * cov - lib/stdin.js
 * Copyright(c) 2013 - 2014 fengmk2 <fengmk2@gmail.com> (http://fengmk2.github.com)
 * MIT Licensed
 */

"use strict";

/**
 * Module dependencies.
 */

module.exports = function stdin(threshold, client, root, pkg) {
  var chunks = [];

  process.stdin.on('data', function (data) {
    chunks.push(data);
  });

  process.stdin.on('end', function () {
    var content = Buffer.concat(chunks).toString(); // coverage.html
    // <div class="percentage">99%</div><div class="sloc">368</div><div class="hits">366</div><div class="misses">2</div>
    var sloc = 0;
    var hits = 0;
    var cov = 0;
    var m = content.match(/\"sloc\">(\d+)<\/div>/);
    if (m) {
      sloc = Number(m[1]) || 0;
    }
    m = content.match(/\"hits\">(\d+)<\/div>/);
    if (m) {
      hits = Number(m[1]) || 0;
    }
    m = content.match(/\"percentage\">([\d\.]+)%?<\/div>/);
    if (m) {
      cov = Number(m[1]) || 0;
    }

    // output:
    // CODE COVERAGE RESULT WAS SAVED TO: http://xxx/xxx.html
    // CODE COVERAGE RESULT OF LINES IS: 157/164

    var pass = cov >= threshold;

    console.log('CODE COVERAGE RESULT OF LINES IS: %s/%s', hits, sloc);
    console.log('CODE COVERAGE RESULT: %s%', cov);
    if (!pass) {
      console.log("CODE COVERAGE BELOW THRESHOLD: %s% < %s%", cov, threshold);
    } else {
      console.log('CODE COVERAGE SUCCEEDED');
    }

    var key = 'cov/html/node-' + process.version + '/'
      + pkg.name + '/' + pkg.version + '.html';
    client.delete(key, function () {
      client.upload(new Buffer(content), {key: key}, function (err, info) {
        if (err) {
          throw err;
        }
        console.log('CODE COVERAGE RESULT WAS SAVED TO: %s', info.url);
        process.exit(pass ? 0 : 1);
      });
    });
  });
};


lib/istanbul.js
/**!
 * cov - lib/istanbul.js
 * Copyright(c) 2014 fengmk2 <fengmk2@gmail.com> (http://fengmk2.github.com)
 * MIT Licensed
 */

"use strict";

/**
 * Module dependencies.
 */

var fs = require('fs');
var path = require('path');
var istanbul = require('istanbul');

module.exports = function istanbulDir(threshold, client, root, pkg) {
  var collector = new istanbul.Collector();
  collector.add(JSON.parse(fs.readFileSync(path.join(root, 'coverage.json'), 'utf8')));
  var actuals = istanbul.utils.summarizeCoverage(collector.getFinalCoverage());
  // console.log(actuals)
  // { lines: { total: 17, covered: 17, skipped: 0, pct: 100 },
  // statements: { total: 17, covered: 17, skipped: 0, pct: 100 },
  // functions: { total: 4, covered: 4, skipped: 0, pct: 100 },
  // branches: { total: 14, covered: 14, skipped: 0, pct: 100 } }

  // 行覆盖率信息: CODE COVERAGE RESULT OF LINES IS: xx/xx
  // 分支覆盖率信息: CODE COVERAGE RESULT OF BRANCHES IS: xx/xx
  // 覆盖率报表链接: CODE COVERAGE RESULT WAS SAVED TO: http://10.32.20.131/t.html

  // output:
  // CODE COVERAGE RESULT WAS SAVED TO: http://xxx/xxx.html
  // CODE COVERAGE RESULT OF LINES IS: 157/164

  var cov = actuals.lines.pct;
  var pass = cov >= threshold;

  // First line must be empty, otherwise mocha analyze will fails on CISE(TOAST) ci env
  console.log('\nCODE COVERAGE RESULT OF LINES IS: %s/%s', actuals.lines.covered, actuals.lines.total);
  console.log('CODE COVERAGE RESULT OF BRANCHES IS: %s/%s', actuals.branches.covered, actuals.branches.total);
  console.log('CODE COVERAGE RESULT OF FUNCTIONS IS: %s/%s', actuals.functions.covered, actuals.functions.total);
  console.log('CODE COVERAGE RESULT OF STATEMENTS IS: %s/%s', actuals.statements.covered, actuals.statements.total);

  console.log('CODE COVERAGE RESULT: %s%', cov);
  if (!pass) {
    console.log("CODE COVERAGE BELOW THRESHOLD: %s% < %s%", cov, threshold);
  } else {
    console.log('CODE COVERAGE SUCCEEDED');
  }

  var lcovDir = path.join(root, 'lcov-report');
  var files = [
    // index.html,
    // tbsession-hsf/index.html
    // tbsession-hsf/lib/index.html
    // tbsession-hsf/lib/tbsession.js.html
  ];
  var walk = function (p) {
    var names = fs.readdirSync(p);
    names.forEach(function (n) {
      var filepath = path.join(p, n);
      if (fs.statSync(filepath).isDirectory()) {
        walk(filepath);
      } else {
        files.push(filepath);
      }
    });
  };

  walk(lcovDir);

  var done = 0;
  var indexURL;
  var version = pkg.version || '0.0.0';
  files.forEach(function (filename, i) {
    var name = 'cov/html/node-' + process.version + '/'
      + pkg.name + '/' + pkg.version + filename.replace(lcovDir, '');
    client.delete(name, function (err) {
      client.uploadFile(filename, {key: name}, function (err, info) {
        done++;
        if (err) {
          console.log('Save %s error: %s', name, err);
        }
        if (filename && filename.indexOf('/lcov-report/index.html') >= 0) {
          indexURL = info.url;
        }
        if (done === files.length) {
          console.log('CODE COVERAGE RESULT WAS SAVED TO: %s', indexURL);
          process.exit(pass ? 0 : 1);
        }
      });
    });
  });
};
